// mpc-js-core/lib/commands/connection.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ConnectionCommands = (function () {
    function ConnectionCommands(protocol) {
        this.protocol = protocol;
    }
    /**
     * Closes the connection to MPD. MPD will try to send the remaining output buffer before it
     * actually closes the connection, but that cannot be guaranteed. This command will not
     * generate a response.
     */
    ConnectionCommands.prototype.close = function () {
        this.protocol.sendCommand('close');
    };
    /**
     * Kills MPD.
     */
    ConnectionCommands.prototype.kill = function () {
        this.protocol.sendCommand('kill');
    };
    /**
     * This is used for authentication with the server. `password` is simply the plaintext password.
     */
    ConnectionCommands.prototype.password = function (password) {
        var cmd = "password \"" + password + "\"";
        return this.protocol.sendCommand(cmd).then(function () { });
    };
    /**
     * Does nothing but return "OK".
     */
    ConnectionCommands.prototype.ping = function () {
        return this.protocol.sendCommand('ping').then(function () { });
    };
    return ConnectionCommands;
}());
exports.ConnectionCommands = ConnectionCommands;
