// mpc-js-core/lib/commands/outputDevice.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var outputDevice_1 = require("../objects/outputDevice");
var OutputDeviceCommands = (function () {
    function OutputDeviceCommands(protocol) {
        this.protocol = protocol;
    }
    /**
     * Returns information about all outputs.
     */
    OutputDeviceCommands.prototype.outputs = function () {
        var _this = this;
        return this.protocol.sendCommand('outputs').then(function (lines) { return _this.protocol.parse(lines, ['outputid'], function (valueMap) { return new outputDevice_1.OutputDevice(valueMap); }); });
    };
    /**
     * Turns an output on.
     */
    OutputDeviceCommands.prototype.enableOutput = function (id) {
        var cmd = "enableoutput " + id;
        return this.protocol.sendCommand(cmd).then(function () { });
    };
    /**
     * Turns an output off.
     */
    OutputDeviceCommands.prototype.disableOutput = function (id) {
        var cmd = "disableoutput " + id;
        return this.protocol.sendCommand(cmd).then(function () { });
    };
    /**
     * Turns an output on or off, depending on the current state.
     */
    OutputDeviceCommands.prototype.toggleOutput = function (id) {
        var cmd = "toggleoutput " + id;
        return this.protocol.sendCommand(cmd).then(function () { });
    };
    return OutputDeviceCommands;
}());
exports.OutputDeviceCommands = OutputDeviceCommands;
