// mpc-js-core/lib/commands/status.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var index_1 = require("../objects/index");
var StatusCommands = (function () {
    function StatusCommands(protocol) {
        this.protocol = protocol;
    }
    StatusCommands.prototype.currentSong = function () {
        var _this = this;
        return this.protocol.sendCommand('currentsong').then(function (lines) { return _this.protocol.parse(lines, ['file'], function (valueMap) { return new index_1.PlaylistItem(valueMap); })[0]; });
    };
    StatusCommands.prototype.status = function () {
        var _this = this;
        return this.protocol.sendCommand('status').then(function (msg) { return _this.protocol.parse(msg, [], function (valueMap) { return new index_1.Status(valueMap); })[0]; });
    };
    StatusCommands.prototype.statistics = function () {
        var _this = this;
        return this.protocol.sendCommand('stats').then(function (msg) { return _this.protocol.parse(msg, [], function (valueMap) { return new index_1.Statistics(valueMap); })[0]; });
    };
    /**
     * Clears the current error message in status
     */
    StatusCommands.prototype.clearError = function () {
        return this.protocol.sendCommand('clearerror').then(function () { });
    };
    return StatusCommands;
}());
exports.StatusCommands = StatusCommands;
