// mpc-js-core/lib/objects/playlists.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var StoredPlaylist = (function () {
    function StoredPlaylist(valueMap) {
        this.name = valueMap.get('playlist');
        this.lastModified = new Date(valueMap.get('Last-Modified'));
    }
    return StoredPlaylist;
}());
exports.StoredPlaylist = StoredPlaylist;
var PlaylistItem = (function () {
    function PlaylistItem(valueMap) {
        this.id = Number(valueMap.get('Id'));
        this.position = Number(valueMap.get('Pos'));
        this.title = valueMap.get('Title');
        this.name = valueMap.get('Name');
        this.artist = valueMap.get('Artist');
        this.album = valueMap.get('Album');
        this.albumArtist = valueMap.get('AlbumArtist');
        this.date = valueMap.get('Date');
        this.genre = valueMap.get('Genre');
        this.duration = Number(valueMap.get('Time'));
        this.path = valueMap.get('file');
        this.lastModified = new Date(valueMap.get('Last-Modified'));
    }
    return PlaylistItem;
}());
exports.PlaylistItem = PlaylistItem;
var SongIdAndPosition = (function () {
    function SongIdAndPosition(valueMap) {
        this.songId = Number(valueMap.get('Id'));
        this.position = Number(valueMap.get('cpos'));
    }
    return SongIdAndPosition;
}());
exports.SongIdAndPosition = SongIdAndPosition;
