// mpc-js-core/lib/objects/status.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * The current status of the player
 */
var Status = (function () {
    function Status(valueMap) {
        this.state = valueMap.get('state');
        this.song = Number(valueMap.get('song'));
        this.songId = Number(valueMap.get('songid'));
        this.nextSong = Number(valueMap.get('nextsong'));
        this.nextSongId = Number(valueMap.get('nextsongid'));
        this.elapsed = Number(valueMap.get('elapsed'));
        this.duration = Number(valueMap.get('duration'));
        this.bitRate = Number(valueMap.get('bitrate'));
        this.audio = valueMap.get('audio');
        var splitAudio = this.audio ? this.audio.split(':') : [];
        this.sampleRate = Number(splitAudio[0]) || undefined;
        this.bitDepth = Number(splitAudio[1]) || undefined;
        this.channels = Number(splitAudio[2]) || undefined;
        this.volume = Number(valueMap.get('volume'));
        this.xfade = Number(valueMap.get('xfade'));
        this.mixrampdb = Number(valueMap.get('mixrampdb'));
        this.mixrampDelay = Number(valueMap.get('mixrampdelay'));
        this.playlistVersion = Number(valueMap.get('playlist'));
        this.playlistLength = Number(valueMap.get('playlistlength'));
        this.repeat = Boolean(Number(valueMap.get('repeat')));
        this.random = Boolean(Number(valueMap.get('random')));
        this.single = Boolean(Number(valueMap.get('single')));
        this.consume = Boolean(Number(valueMap.get('consume')));
        this.updating = Number(valueMap.get('updating_db'));
        this.error = valueMap.get('error');
    }
    return Status;
}());
exports.Status = Status;
